File: src/main/java/com/owncloud/android/utils/FileSortOrder.java
/*
 * Nextcloud Android client application
 *
 * @author Sven R. Kunze
 * Copyright (C) 2017 Sven R. Kunze
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU AFFERO GENERAL PUBLIC LICENSE
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU AFFERO GENERAL PUBLIC LICENSE for more details.
 *
 * You should have received a copy of the GNU Affero General Public
 * License along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.owncloud.android.utils;

import com.owncloud.android.datamodel.OCFile;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * Sort order 
 */

public class FileSortOrder {
    public static final FileSortOrder sort_a_to_z = new FileSortOrderByName("sort_a_to_z", true);
    public static final FileSortOrder sort_z_to_a = new FileSortOrderByName("sort_z_to_a", false);
    public static final FileSortOrder sort_old_to_new = new FileSortOrderByDate("sort_old_to_new", true);
    public static final FileSortOrder sort_new_to_old = new FileSortOrderByDate("sort_new_to_old", false);
    public static final FileSortOrder sort_small_to_big = new FileSortOrderBySize("sort_small_to_big", true);
    public static final FileSortOrder sort_big_to_small = new FileSortOrderBySize("sort_big_to_small", false);

    public static final Map<String, FileSortOrder> sortOrders;
    static {
        sortOrders = new HashMap<String, FileSortOrder>();
        sortOrders.put(sort_a_to_z.mName, sort_a_to_z);
        sortOrders.put(sort_z_to_a.mName, sort_z_to_a);
        sortOrders.put(sort_old_to_new.mName, sort_old_to_new);
        sortOrders.put(sort_new_to_old.mName, sort_new_to_old);
        sortOrders.put(sort_small_to_big.mName, sort_small_to_big);
        sortOrders.put(sort_big_to_small.mName, sort_big_to_small);
    }

    public String mName;
    public boolean mAscending;

    public FileSortOrder(String name, boolean ascending) {
        mName = name;
        mAscending = ascending;
    }

    public ArrayList<OCFile> sortCloudFiles(ArrayList<OCFile> files) {
        return sortCloudFilesByFavourite(files);
    }

    public ArrayList<File> sortLocalFiles(ArrayList<File> files) {
        return files;
    }

    /**
     * Sorts list by Favourites.
     *
     * @param files files to sort
     */
    public static ArrayList<OCFile> sortCloudFilesByFavourite(ArrayList<OCFile> files) {
        Collections.sort(files, (o1, o2) -> {
            if (o1.getIsFavorite() && o2.getIsFavorite()) {
                return 0;
            } else if (o1.getIsFavorite()) {
                return -1;
            } else if (o2.getIsFavorite()) {
                return 1;
            }
            return 0;
        });

        return files;
    }
}

File: app/src/main/java/com/owncloud/android/utils/FileSortOrder.kt
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2022 Álvaro Brey <alvaro@alvarobrey.com>
 * SPDX-FileCopyrightText: 2017 Sven R. Kunze <srkunze@mail.de>
 * SPDX-License-Identifier: AGPL-3.0-or-later OR GPL-2.0-only
 */
package com.owncloud.android.utils

import com.owncloud.android.datamodel.OCFile
import com.owncloud.android.lib.resources.trashbin.model.TrashbinFile
import java.io.File
import java.util.Collections

/**
 * Sort order
 */
open class FileSortOrder(@JvmField var name: String, var isAscending: Boolean) {

    val sortMultiplier: Int
        get() = if (isAscending) 1 else -1

    @Suppress("EnumNaming", "EnumEntryName") // already saved in user preferences -.-'
    enum class Type {
        trashBinView,

        @Suppress("ktlint:standard:enum-entry-name-case")
        localFileListView
    }

    enum class SortType {
        SIZE,
        ALPHABET,
        DATE
    }

    companion object {
        const val SORT_A_TO_Z_ID = "sort_a_to_z"
        const val SORT_Z_TO_A_ID = "sort_z_to_a"
        const val SORT_OLD_TO_NEW_ID = "sort_old_to_new"
        const val SORT_NEW_TO_OLD_ID = "sort_new_to_old"
        const val SORT_SMALL_TO_BIG_ID = "sort_small_to_big"
        const val SORT_BIG_TO_SMALL_ID = "sort_big_to_small"

        @JvmField
        val SORT_A_TO_Z: FileSortOrder = FileSortOrderByName(SORT_A_TO_Z_ID, true)

        @JvmField
        val SORT_Z_TO_A: FileSortOrder = FileSortOrderByName(SORT_Z_TO_A_ID, false)

        @JvmField
        val SORT_OLD_TO_NEW: FileSortOrder = FileSortOrderByDate(SORT_OLD_TO_NEW_ID, true)

        @JvmField
        val SORT_NEW_TO_OLD: FileSortOrder = FileSortOrderByDate(SORT_NEW_TO_OLD_ID, false)

        @JvmField
        val SORT_SMALL_TO_BIG: FileSortOrder = FileSortOrderBySize(SORT_SMALL_TO_BIG_ID, true)

        @JvmField
        val SORT_BIG_TO_SMALL: FileSortOrder = FileSortOrderBySize(SORT_BIG_TO_SMALL_ID, false)

        @JvmField
        val sortOrders: Map<String, FileSortOrder> = Collections.unmodifiableMap(
            mapOf(
                SORT_A_TO_Z.name to SORT_A_TO_Z,
                SORT_Z_TO_A.name to SORT_Z_TO_A,
                SORT_OLD_TO_NEW.name to SORT_OLD_TO_NEW,
                SORT_NEW_TO_OLD.name to SORT_NEW_TO_OLD,
                SORT_SMALL_TO_BIG.name to SORT_SMALL_TO_BIG,
                SORT_BIG_TO_SMALL.name to SORT_BIG_TO_SMALL
            )
        )

        /**
         * Sorts list by Favourites.
         *
         * @param files files to sort
         */
        @JvmStatic
        fun sortCloudFilesByFavourite(files: MutableList<OCFile>): List<OCFile> {
            files.sortWith { o1: OCFile, o2: OCFile ->
                when {
                    o1.isFavorite && o2.isFavorite -> 0
                    o1.isFavorite -> -1
                    o2.isFavorite -> 1
                    else -> 0
                }
            }
            return files
        }
    }

    open fun sortCloudFiles(files: MutableList<OCFile>): List<OCFile> {
        return sortCloudFilesByFavourite(files)
    }

    open fun sortLocalFiles(files: MutableList<File>): List<File> {
        return files
    }

    open fun sortTrashbinFiles(files: MutableList<TrashbinFile>): List<TrashbinFile> {
        return files
    }

    open fun getType(): SortType {
        return when (name) {
            SORT_Z_TO_A_ID,
            SORT_A_TO_Z_ID -> SortType.ALPHABET

            SORT_SMALL_TO_BIG_ID,
            SORT_BIG_TO_SMALL_ID -> SortType.SIZE

            SORT_NEW_TO_OLD_ID,
            SORT_OLD_TO_NEW_ID -> SortType.DATE

            else -> SortType.ALPHABET
        }
    }
}
